topics/tutorials/file_and_code_templates/using_file_templates.md

```diff
@@ -0,0 +1,131 @@
+It allows customizing the action by providing an action name, icon, overriding methods creating dialog, and others.
+Example action:
+
+```java
+public class CreateMyClassAction extends CreateFileFromTemplateAction {
+
+  public CreateMyClassAction() {
+    super("My Class", "Creates new class", CLASS_ICON);
+  }
+
+  @Override
+  protected void buildDialog(Project project, PsiDirectory directory,
+      CreateFileFromTemplateDialog.Builder builder) {
+    builder
+    .setTitle("New My File")
+    .addKind("Class", CLASS_ICON, "My Class");
+  }
+
+  @Override
+  protected String getActionName(PsiDirectory directory,
+      @NotNull String newName, String templateName) {
+    return "Create My Class: " + newName;
+  }
+}
+```
+
+The new action should be registered under the `NewGroup` group, e.g:
+
+```xml
+<actions>
+  <action id="Create.MyClass" class="com.example.CreateMyClassAction">
+    <add-to-group group-id="NewGroup"/>
+  </action>
+</actions>
+```
+
+### Custom "Create File From Template" Actions
+
+In some cases, the default mechanism for creating files from templates is insufficient.
+Consider a language that defines multiple types of core entities, e.g., in the Java language, the following entities can be created:
+* Class
+* Interface
+* Record
+* Enum
+* Annotation
+
+Having all of those items in the <menupath>File | New</menupath> action group may overwhelm users with the number of options to choose.
+It is more user-friendly to provide a single <menupath>File | New | Java Class</menupath> action and let users choose a specific entity type in the creation dialog:
+
+![Create Java class](new_java_class_popup.png){width="330"}
+
+This can be achieved by placing templates in the _[Internal](providing_file_templates.md#internal)_ category, so they are not picked up by the default mechanism, and then registering a custom
+[`CreateFileFromTemplateAction`](upsource:///platform/lang-impl/src/com/intellij/ide/actions/CreateFileFromTemplateAction.java)
+implementation as described in the [](#exposing-file-templates-from-the-other-category) section.
+To provide a list of selectable entity types, the action should provide a custom dialog with multiple file kinds for every template type, e.g.:
+
+```java
+@Override
+protected void buildDialog(Project project, PsiDirectory directory,
+    CreateFileFromTemplateDialog.Builder builder) {
+  builder
+    .setTitle("My File")
+    .addKind("Class", CLASS_ICON, "My Class")
+    .addKind("Record", RECORD_ICON, "My Record")
+    .addKind("Enum", ENUM_ICON, "My Enum");
+}
+```
+
+As file templates are placed in the <path>fileTemplates/internal</path> directory, they are not listed in the <menupath>Settings/Preferences | Editor | File and Code Templates</menupath> settings page, and users can't adjust them to their needs.
+Internal templates can be exposed in the _Files_ category by additionally registering them via the `com.intellij.internalFileTemplate` EP, e.g.:
```

Comment:
control tag

Reply:
Fixed, also in all other places. Thanks.